[memory_model] import Bytes trait from rust-vmm/vm-memory

serban300 opened the pull request:
## Reason for This PR

implements #1470 

## Description of Changes

Provided in the commits description.

## License Acceptance

By submitting this pull request, I confirm that my contribution is made under
the terms of the Apache 2.0 license.

## PR Checklist

`[Author TODO: Meet these criteria. Where there are two options, keep one.]`
`[Reviewer TODO: Verify that these criteria are met. Request changes if not]`

- [x] All commits in this PR are signed (`git commit -s`).
- [x] This PR is linked to an issue.
- [x] The description of changes is clear and encompassing.
- [x] No docs need to be updated as part of this PR.
- [x] Code-level documentation for touched code is included in this PR.
- [x] No API changes are included in this PR.
- [x] The changes in this PR have no user impact.
- [x] No new `unsafe` code has been added.

aghecenco commented on src/memory_model/src/guest_memory.rs:
```diff
@@ -505,6 +477,14 @@ impl GuestMemory {
 impl Bytes<GuestAddress> for GuestMemory {
     type E = Error;
+
+    fn write_slice(&self, buf: &[u8], addr: GuestAddress) -> std::result::Result<usize, Self::E> {
```
Can you add the docs and example comments? The function shows up in the generated documentation.

serban300 replied:
Done.

aghecenco commented on src/memory_model/src/guest_memory.rs:
```diff
@@ -485,6 +456,18 @@ impl Bytes<GuestAddress> for GuestMemory {
         })
     }
+
+    fn read_slice(
```
Same here for the documentation and examples comments

serban300 replied:
Done.

aghecenco commented on src/memory_model/src/guest_memory.rs:
```diff
@@ -468,6 +435,14 @@ impl Bytes<GuestAddress> for GuestMemory {
         })
     }
+
+    fn read_obj<T: ByteValued>(&self, addr: GuestAddress) -> std::result::Result<T, Self::E> {
```
Same here for documentation and example comments

serban300 replied:
Done.

serban300 commented on src/memory_model/src/bytes.rs:
```diff
@@ -48,3 +50,15 @@ byte_valued_type!(i16);
+///
+/// Candidates which may implement this trait include:
+/// - anonymous memory areas
+/// - mmapped memory areas
```
Done.

serban300 commented on src/memory_model/src/bytes.rs:
```diff
@@ -61,4 +59,10 @@ byte_valued_type!(isize);
+    /// Writes the entire contents of a slice into the container at the specified address.
+    ///
+    /// Returns an error if there isn't enough room within the container to complete the entire
+    /// write. Part of the data may have been written nevertheless.
```
Yes. Exactly.

serban300 commented:
> Had some pending nits before the `push -f` smile The `memory_model` functions that have been migrated to the `Bytes` trait should still be documented with usage examples, they show up in the generated documentation. Besides, we already have the comments, they just need to be adapted.

Good point. Done.

andreeaflorescu commented on src/memory_model/src/bytes.rs:
```diff
@@ -48,3 +50,59 @@ byte_valued_type!(i16);
+    /// * `addr` - Begin writing at this address.
+    /// * `src` - Copy from `src` into the container.
+    /// * `count` - Copy `count` bytes from `src` into the container.
+    fn read_from<F>(&self, addr: A, src: &mut F, count: usize) -> Result<usize, Self::E>
```
I am really unhappy with the naming from vm-memory in this case. The functions should be named from the memory perspective. You will call the read_from function on a guest_memory object.

The code looks like this:
```rust
let guest_memory = ...;
guest_memory.read_from(addr, my_file, count);
```

When I read the code `guest_memory.read_from` I assume that I read from Guest Memory because the operation is applied on a guest memory object and so the operation refers to the guest memory. This is actually write from the guest memory point of view. The operation you apply on the guest memory is to write in it the contents of a something like a File.

I would also open an issue in vm-memory about this if you don't have anything against it.

serban300 replied:
I will think about this and maybe try to find some better names, but I would suggest merging this PR first and I'll submit another PR after changing the method names in rust-vmm/vm-memory crate.

bonzini replied:
Naming things is hard :( but note that vm-memory isn't the first crate that adopts this naming, see [here](https://docs.rs/imagefmt/4.0.0/imagefmt/fn.read_from.html) or [here](https://doc.servo.org/tendril/stream/trait.TendrilSink.html). The cue here *should* come from the preposition:

* `x.read()` translates to "hey `x`, read something!", so the effect is to read from `x` (and get back what's read).

* instead, `x.read_from(y)` translates to "hey `x`, read something from `y`!", so it reads from `y` (and places what's read into `x`).

andreeaflorescu replied:
@bonzini I will take a look at the crates you mentioned. I would still look for a better way to express the operations we are doing on the guest memory if you don't mind.

serban300 commented:
@aghecenco @andreeaflorescu I addressed all the comments. Please take another look !

andreeaflorescu commented on src/devices/src/virtio/net.rs:
```diff
@@ -229,9 +229,9 @@ impl NetEpollHandler {
-                        Ok(sz) => {
+                        Ok(()) => {
                             METRICS.net.rx_count.inc();
-                            write_count += sz;
```
With this change the behavior of this function changes. Before we were not returning an error if only a part of the slice was successfully written. Probably here we want to check if the Error is PartialBuffer and get the number of actual bytes written; then we need to increment the write count.

serban300 replied:
Done.

andreeaflorescu commented on src/devices/src/virtio/net.rs:
```diff
@@ -447,8 +447,8 @@ impl NetEpollHandler {
                     desc_addr,
                 );
                 match read_result {
-                    Ok(sz) => {
```
Same comment as for write_result.

serban300 replied:
Done.

serban300 commented on src/arch/src/aarch64/fdt.rs:
```diff
@@ -129,12 +129,9 @@ pub fn create_fdt<T: DeviceInfoForFDT + Clone + Debug>(
         .write_slice(fdt_final.as_slice(), fdt_address)
         .map_err(Error::WriteFDTToMemory)?;
-    if written < FDT_MAX_SIZE {
-        return Err(Error::IncompleteFDTMemoryWrite);
```
I removed `IncompleteFDTMemoryWrite`.

I don't think such a method exists. At least not that I know of. And for the moment I don't know of any scenario in which we would need to retry. 

Probably the idea is that  `write_slice` doesn't need to return the number of written bytes. The number of written bytes is `buf.len()` in case of success.

andreeaflorescu replied:
Yeah, I just realized afterwards that you are suppose to get the number of written bytes from the error, which I find a bit weird tbh.

bonzini replied:
If it's expected to read only a part of the slice, it is also possible to use `mem.read_from` since slices implement `Read`. Usually it's not expected (as is this case) and vm-memory's semantics allows using `?` more effectively.

serban300 commented:
@aghecenco @andreeaflorescu please take another look. I addressed all the comments.

I had to decrease the coverage by 0.1% because I added some lines that can't be tested. For example [here](https://github.com/serban300/firecracker/blob/5b522c91e53e446894d342403833a453e6fec952/src/devices/src/virtio/net.rs#L240) .

bonzini commented:
> For example here .

That line should be covered by

    // We make the prev desc write_only (with no other flag) to get a chain which is
    // writable, but too short.

Perhaps you're thinking of the tx PartialBuffer case, which is indeed not covered (you probably could test it by creating a descriptor that ends after the end of memory).
